src/conv-core/cpuaffinity.C

```diff
@@ -442,62 +453,226 @@ static void cpuAffinityHandler(void *m)
 */
   CmiAssert(CmiMyPe()==0 && rankmsg != NULL);
   skt_ip_t & ip = msg->ip;
-  pe = msg->pe;
+  int node = msg->node;
+  int pe = msg->pe;
   auto iter = hostTable.find(ip);
   if (iter != hostTable.end()) {
     rec = iter->second;
     CmiFree(msg);
   }
   else {
     rec = msg;
-    rec->seq = nodecount;
-    nodecount++;                          /* a new node record */
+    rec->seq = hostcount++;
     hostTable.emplace(ip, msg);
   }
-  myrank = rec->rank%rec->ncores;
-  while (in_exclude(myrank)) {             /* skip excluded core */
-    myrank = (myrank+1)%rec->ncores;
-    rec->rank ++;
+
+  int & node_host = rankmsg->node_hosts[node];
+  if (node_host == -1)
+  {
+    rankmsg->node_ranks_on_host[node] = rec->node_rank++;
+    node_host = rec->seq;
   }
-  rankmsg->ranks[pe] = myrank;             /* core rank */
-  rankmsg->nodes[pe] = rec->seq;           /* on which node */
-  rec->rank ++;
-  count ++;
-  if (count == CmiNumPes()) {
-    DEBUGP(("Cpuaffinity> %zu unique compute nodes detected!\n", hostTable.size()));
+  rankmsg->node_numpes[node]++;
+  rankmsg->pe_ranks_on_host[pe] = rec->pe_rank++;
+  rankmsg->pe_hosts[pe] = rec->seq;
+
+  if (++count == npes) {
+    DEBUGP(("Cpuaffinity> %zu unique hosts detected!\n", hostTable.size()));
     for (const auto & pair : hostTable)
       CmiFree(pair.second);
     hostTable.clear();
-#if 1
+
     /* bubble sort ranks on each node according to the PE number */
-    {
     int i,j;
+    for (i=0; i<nnodes-1; i++)
```

Comment:
Another nitpick, but unless I'm missing something, I think this loop and the below loop can both be done in linear time by iterating through and directly assigning values in ascending order per node/PE on the same host rather than doing this.

Reply:
Could you share some code or pseudocode to illustrate what you mean?

Reply:
I've implemented it in #2672.